### ISSUE TITLE
help: using pytest with shellcmd power control 

### ISSUE THREAD

**Issue by erdemkahraman** [siemens/mtda#381]
Hi everybody, I am playing with mtda (0.22) and trying to understand how it is working.

According to my old experiences, I am able to use this tool with console e.g. "mtda-cli target on" and It is working well..

Also here is my configuration;
```
debug=3
[console]
variant=serial
port=/dev/ttyUSB0
rate=115200
[storage]
variant=usbf
file=/dev/mmcblk1p3
[power]
variant=shellcmd
on-cmd=/opt/power.sh on
off-cmd=/opt/power.sh off
check-on=/opt/power.sh check
```

But on my pytest environment I guess mtda is not able to reach the /etc/mtda/config.

Here is my test funcs.

```
class HARDWARE(Device):
    def __init__(self, disk_path):
        self.disk_path = disk_path


    def hardware_setup(self):
        subprocess.call(["systemctl", "restart", "mtda"])
        subprocess.call(["cat", "/etc/mtda/config"])
        Test.setup()
        assert Target.off() is True
        assert Storage.to_host() is True
        Storage.write(self.disk_path)
        assert Storage.to_target() is True
        assert Target.on() is True
        Test.teardown()
    
    def setup(self):
        self.hardware_setup()

    def teardown(self):
        Test.setup()
        assert Target.off() is True
        Test.teardown()
```

It is calling this hardware_setup function on my test. Here is the logs about this code.

```
>          device.setup()
tests/testsuites/conftest.py:31: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
self = <tests.devices.HARDWARE object at 0x7f81f4af40>
    def setup(self):
>       self.hardware_setup()
tests/devices.py:288: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
self = <tests.devices.HARDWARE object at 0x7f81f4af40>
    def hardware_setup(self):
        subprocess.call(["systemctl", "restart", "mtda"])
        subprocess.call(["cat", "/etc/mtda/config"])
>       Test.setup()
tests/devices.py:279: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
    def setup():
>       assert Test.init() is True
/app/hw-test-env/lib/python3.8/site-packages/mtda/pytest.py:59: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
    def init():
        if pytest.mtda is not None:
            return True
    
        # Initialize client
        pytest.logging = False
        pytest.mtda = mtda.client.Client()
        pytest.output = TestOutput()
        remote = pytest.mtda.remote()
        pytest.mtda.console_remote(remote, pytest.output)
        pytest.mtda.monitor_remote(remote, pytest.output)
        pytest.mtda.start()
    
        # Start test-suite with the DUT powered off
>       assert Target.off() is True
E       AssertionError
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
self = <tests.devices.HARDWARE object at 0x7f81f4af40>
    def teardown(self):
        Test.setup()
>       assert Target.off() is True
E       AssertionError
tests/devices.py:295: AssertionError
-------------------------------------------------------------------
# debug3: main.load_timeouts_config()
# debug3: main.load_timeouts_config: None
# debug3: main.console_remote()
# debug3: main.console_remote(): None
# debug3: main.monitor_remote()
# debug3: main.monitor_remote(): None
# debug3: main.start()
# debug3: main.target_status()
# debug3: main._target_status()
# debug3: main._target_status(): ???
# debug3: main.target_status(): ???
# debug3: main.target_off()
# debug3: main._session_check(Shiva)
# debug3: session Shiva: 300 seconds
# debug3: main._session_event(ACTIVE Shiva)
# debug3: main.notify(SESSION,ACTIVE Shiva)
# debug3: main.notify: None
# debug3: main._session_event: None
# debug3: main._session_check: None
# debug3: main._target_status()
# debug3: main._target_status(): ???
# debug3: main.power_locked()
# debug3: main._session_check(Shiva)
# debug3: session Shiva: 300 seconds
# debug3: main._session_check: None
# debug3: main.power_locked(): True
# debug3: main.target_off(): False
# debug3: main.console_clear()
# debug3: main._session_check(Shiva)
# debug3: session Shiva: 300 seconds
# debug3: main._session_check: None
# debug3: main.console_locked()
# debug3: main._session_check(Shiva)
# debug3: session Shiva: 300 seconds
# debug3: main._session_check: None
# debug3: main._check_locked()
# debug3: main.target_owner()
# debug3: main.console_locked(): False
# debug3: main.console_clear(): None
# debug3: main.config_set_power_timeout()
# debug3: main._session_check(None)
# debug3: session Shiva: 299 seconds
# debug3: main._session_check: None
# debug3: main.config_set_power_timeout(): 3600
# debug3: main.config_set_session_timeout()
# debug3: main._session_check(None)
# debug3: session Shiva: 9 seconds
# debug3: main._session_check: None
# debug3: main.target_status()
# debug3: main._target_status()
# debug3: main._target_status(): ???
# debug3: main.target_status(): ???
# debug3: main.target_off()
# debug3: main._session_check(Shiva)
# debug3: session Shiva: 10 seconds
# debug3: main._session_check: None
# debug3: main._target_status()
# debug3: main._target_status(): ???
# debug3: main.power_locked()
# debug3: main._session_check(Shiva)
# debug3: session Shiva: 10 seconds
# debug3: main._session_check: None
# debug3: main.power_locked(): True
# debug3: main.target_off(): False
```

Also output of the cat /etc/mtda/config on pytest test case; It is able to print it to the console.
Why this test is not able to get the target status? Is there any missing on my mind? Is there any tricky way?

**Comment by chombourger** [comment]
logs provided above appear to be client-side logs. people may want to see server-side logs (i.e. `journalctl -u mtda -b`)
note that restarting the `mtda` service from your pytest code is rather dubious

it is also recommended to run the client and server on different nodes where the server is usually a Pi device of some sort, power of the DUT controlled using a relay

**Comment by erdemkahraman** [comment]
I am running a docker container; and mtda's test is running inside of docker container. 
I have to do this test on gitlab-runner's docker executor and I've binded the systemd volumes to the docker container;
That's why MTDA is able to reach the host's mtda service from container. I guess this way (mtda is inside docker container) is illegal but I am challenging.

It was working when I am using the console client tool : mtda-cli 
But I could not get the problem on pytest environment.

Also these client logs are not appearing on the host's journal logs. It was appearing when I am using mtda-cli tool.

**Comment by erdemkahraman** [comment]
more logs are here

```
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
self = <tests.devices.HARDWARE object at 0x7f884fbb50>
    def setup(self):
>       self.hardware_setup()
tests/devices.py:286: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
self = <tests.devices.HARDWARE object at 0x7f884fbb50>
    def hardware_setup(self):
        subprocess.call(["cat", "/etc/mtda/config"])
>       assert Target.off() is True
tests/devices.py:278: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
    def off():
>       pytest.output.clear_power()
/app/hw-test-env/lib/python3.8/site-packages/mtda/pytest.py:206: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
name = 'output'
    def __getattr__(name: str) -> object:
        if name == "Instance":
            # The import emits a deprecation warning.
            from _pytest.python import Instance
    
            return Instance
>       raise AttributeError(f"module {__name__} has no attribute {name}")
E       AttributeError: module pytest has no attribute output
/app/hw-test-env/lib/python3.8/site-packages/pytest/__init__.py:169: AttributeError
During handling of the above exception, another exception occurred:
```

**Comment by erdemkahraman** [comment]
Nevermind my docker container case, I am trying to run this sample tests on my Ubuntu 22.04 machine they are giving same error. Yes I am using shellcmd as powercontroller. Pytest logs are here;
[pytest-mtda.log](https://github.com/siemens/mtda/files/13732463/pytest-mtda.log)

**Comment by erdemkahraman** [comment]
Also I've tried it with the gpio on power option;

conf:
```
[main]
debug=3

[power]
variant=gpio
gpio=gpiochip0@3
```

I have tried it by using ```mtda-cli target on```; It working properly.
Then I have ran a test on mtda directory like;
```q@raspberrypi:~/mtda/tests$ python3 -m pytest -x --setup-show --verbose -rA --capture=sys --pyargs -o log_cli=true --log-cli-level=DEBUG --capture=tee-sys test_target```

But the result is same with before; [pytest-mtda.log](https://github.com/siemens/mtda/files/13732463/pytest-mtda.log)

I guess I am missing something on running test with pytest, but [documentation](https://mtda.readthedocs.io/en/latest/integration.html#pytest) was simple :)

**Comment by erdemkahraman** [comment]
Hi I've tested the [test-using-docker](https://github.com/siemens/mtda/issues/scripts/test-using-docker) and It has worked properly, 

I have ran a new test using my own configuration with a similar technique in this script and it did not work. My thought is; Could it be that mtda pytest support is not working properly for shellcmd? Has pytest support been tested for a different target other than Docker?
